### PR TITLE
[AIRFLOW-2691] Makes Airflow's JS code (and dependencies) manageable via npm and webpack

### PR DESCRIPTION
**Pull request by verdan**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-2691) issues and references them in the PR title. 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In an effort to simplify the JS dependencies of Airflow, it introduces the `npm` and `webpack` for the package management. For now, it only implements this in the www_rbac version of the web server. 
The problem with the frontend (js) code of Airflow is that most of the custom JS is written in the html files, using the Flask's (Jinja) variables in that JS. The next step of this effort would be to extract that custom .js code in separate JS files, and use the dependencies in those files using `require` or `import`. (At the moment, I'm simply using the CopyWebPackPlugin to copy the required dependencies for use)
There are also some dependencies which are directly modified in the codebase. I couldn't found the npm versions of those libraries. (**dagre-d3.js** and **gantt-chart-d3v2.js**). Apparently dagre-d3.js that we are using is one of the gist or is very old version, while the gantt-chart-d3v2 has been modified according to Airflow's requirements. 
This PR also adds a new command while making a build for distribution i.e., `compile_assets`. The details of the command and usage has also been added in the dev documentation here: https://github.com/verdan/incubator-airflow/blob/b5a975f2afb4e29063ccfb17343ee07e5363cacb/dev/README.md
(Someone need to update the release process documentation on Confluence/Wiki as I don't have access to edit or comment on [this page](https://cwiki.apache.org/confluence/display/AIRFLOW/Releasing+Airflow#ReleasingAirflow-PublishingtoPyPi))

Next Steps: 
- Extract custom JS from html files, and put them in separate JS files. 
- Introduce eslint, and test cases for JS code. 

_NOTE: This webpack/npm setup is pretty much same as Apache's Superset._ 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by verdan** [comment]
@bolkedebruin @Fokko PTAL

**Comment by bolkedebruin** [comment]
Nice one @verdan Can you please raise some attention to this on the mailing list?

File: airflow/www_rbac/static/webpack.config.js
Comment (ashb) [inline airflow/www_rbac/static/webpack.config.js]: Does this file actually need to be in the `static/` folder -- is it needed when the webserver is running, or just in the JS built stage?

File: airflow/www_rbac/templates/airflow/conn_create.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/conn_create.html:22]: Why the change of underscore to camelCase? There doesn't seem to be a connectionForm.j either?

File: airflow/www_rbac/templates/airflow/dag.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/dag.html:24]: What happened to this?

File: airflow/www_rbac/templates/airflow/dag.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/dag.html:77]: All these extra non-functional changes make this PR _very_ hard to review.

File: airflow/www_rbac/templates/airflow/dags.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/dags.html:229]: New JS file -why's that?

File: airflow/www_rbac/templates/airflow/gantt.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/gantt.html:52]: Again - why the change in style?

File: airflow/www_rbac/templates/airflow/graph.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/graph.html]: Make this a Jinja comment, not an HTML comment -- we don't need this in the output.

File: airflow/www_rbac/templates/airflow/graph.html
Comment (ashb) [inline airflow/www_rbac/templates/airflow/graph.html:25]: What's happened to this CSS file? Is it needed still?

File: airflow/www_rbac/templates/appbuilder/baselayout.html
Comment (ashb) [inline airflow/www_rbac/templates/appbuilder/baselayout.html]: Again, HTML comment should be Jinja comment

File: pypi_push.sh
Comment (ashb) [inline pypi_push.sh]: yarn _and_ npm?!

File: pypi_push.sh
Comment (ashb) [inline pypi_push.sh]: Having this in a script is a good first step - but this is not enough for us. This needs to be integrated into `setup.py` so that running `sdist` does these steps - the "official" releases of Apache projects are onto dist.apache.org. Also having it run directly as part of setup.py means it can't be forgotten.

File: airflow/www_rbac/static/webpack.config.js
Comment (ashb) [inline airflow/www_rbac/static/webpack.config.js]: Ah - this maps into `ganntCharD3v2.$hash.js`. This somewhat breaks the mapping between input filename and output filename - please have these be `'gantt-chart-d3v2': STATIC_DIR + '/js/gantt-chart-d3v2.js'` etc.

**Review by ashb — CHANGES_REQUESTED** [review]
Having webpack is great and will really help tidy up our front end build process.

Two things though:

What is the advantage of moving some of the JS code from inline to a file? For instance the clock snippet/clock.js is basically unchanged.

Secondly: with this change does it mean we can't just checkout the repo and run the (RBAC) webserver? (I'm not sure if that's a problem per se, I just want to check)

File: airflow/www_rbac/static/package.json
Comment (ashb) [inline airflow/www_rbac/static/package.json]: Something about having all these type of files inside `static/` feels wrong to me -- seeing that directory I would expect that just to contain static files to be served by a webserver - not config files such as this

File: airflow/www_rbac/static/webpack.config.js
Comment (ashb) [inline airflow/www_rbac/static/webpack.config.js]: Does this generate source maps? If not please can you enable that - even for "prod" builds please.

File: airflow/www_rbac/static/webpack.config.js
Comment (ashb) [inline airflow/www_rbac/static/webpack.config.js]: > Since webpack v4 the extract-text-webpack-plugin should not be used for css. Use mini-css-extract-plugin instead.

What does this plugin do anyways?

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: This file needs to be in the same folder as the package.json (A configuration file for npm comamnd)
It is needed when you run any of the `npm run prod|build|dev|etc...` commands. (without running these commands you won't get required JS, CSS). I'll add a detailed comment as how the npm and webpack works.

File: airflow/www_rbac/templates/airflow/dag.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/dag.html:24]: This css has already been included in the base file i.e., `appbuilder/baselayout.html` 

**structure:** dag.html `extends > ` airflow/master.html `extends >` appbuilder/baselayout.html

File: airflow/www_rbac/templates/airflow/dag.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/dag.html:77]: Apologies for these non-functional changes. OCD 😢

File: airflow/www_rbac/templates/airflow/dags.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/dags.html:229]: The existing [`jquery.dataTables.min.js`](https://github.com/apache/incubator-airflow/blob/master/airflow/www_rbac/static/jquery.dataTables.min.js) is a combination of jQuery Data Tables and Data Tables Bootstrap (for stylings), which is exactly same as this. They don't ship that in a single build as npm package, so needed to add both.

File: airflow/www_rbac/templates/airflow/graph.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/graph.html]: roger that!!

File: airflow/www_rbac/templates/airflow/graph.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/graph.html:25]: We had 2 .css files included in graph.html. dagre.css and graph.css, which makes no sense as both were serving the same purpose i.e., styling the DAG, so combined the css of both in graph.css now, and removed dagre.css (as graph.css makes more sense in graph.html, keeping the same name)

File: airflow/www_rbac/templates/appbuilder/baselayout.html
Comment (verdan) [reply airflow/www_rbac/templates/appbuilder/baselayout.html]: 👍

File: pypi_push.sh
Comment (verdan) [reply pypi_push.sh]: `yarn` is a replacement of `npm install` here because it is faster and more deterministic when installing packages. We can also use `npm install` but it would be slow.

File: pypi_push.sh
Comment (verdan) [reply pypi_push.sh]: This script generates the frontend build, make an Airflow build and tries to upload it to pypi. So instead of running the `python setup.py sdist upload` separately, we can simply use this script as an automated script to upload the new release. Thoughts? 
P.S: I followed the same pattern as [`incubator-superset`](https://github.com/apache/incubator-superset/blob/master/pypi_push.sh).

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: I agree that it might be confusing for some users, but it has pros and cons. 
To differentiate the local javascript file from the processed javascript file and to eliminate confusion, we usually use different names. 
After every change in javascript file, you'll need to process the javascript file i.e, minimize, format and include dependencies etc. So, if the names in the `<script>` tag in html will be in different format, you'd know that this is a processed file, and you need to make the frontend build again using `npm run build`

Hint: you can simply run `npm run dev` which will constantly watch and automatically makes a build on every javascript change you make in any file.

File: airflow/www_rbac/static/package.json
Comment (verdan) [reply airflow/www_rbac/static/package.json]: This file holds all dependencies (JS packages, and their versions) and the commands to build those packages. I was thinking to place this in `www_rbac` directory, but these are purely frontend (JS & CSS) dependent files, and usually placed under static or assets directories.

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: will enable source maps. good point!!

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: nice catch about the deprecation!!! 
So, if you have some CSS code (hard-coded or imported) in a JS file, this plugin will extract that in a separate `.css or .min.css` file and then we can use that .min.css file in our code.

**Comment by verdan** [comment]
@ashb replying to your questions:

> What is the advantage of moving some of the JS code from inline to a file? For instance the clock snippet/clock.js is basically unchanged.

A couple of advantages of doing that. 
- We'd be able combine our javascript and all the dependencies (like d3.js, nvd3.js) in one single minified chunk, which would be efficient and improve the page load time. 
- Write the test cases of our JS code by using separate files. 

> Secondly: with this change does it mean we can't just checkout the repo and run the (RBAC) webserver? (I'm not sure if that's a problem per se, I just want to check)

As a contributor, if you want to contribute in the repo, you'll need to compile the javascript at least once, using `npm run build` and it will remain in your `static/dist/`. I've also added the details in Contributing guidelines on how to build the JS code. But it would be great to rebuild everytime you pull the lates code from master branch, just to make sure you are completely synced with all the new libraries/packages.
(If you'll make changes in javascript then for every change you'll need to recompile javascript, or use `npm run dev` to automatically compile JS on every change).

File: airflow/www_rbac/templates/airflow/conn_create.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/conn_create.html:22]: https://github.com/apache/incubator-airflow/pull/3572#discussion-diff-200991459R40

File: airflow/www_rbac/templates/airflow/gantt.html
Comment (verdan) [reply airflow/www_rbac/templates/airflow/gantt.html:52]: https://github.com/apache/incubator-airflow/pull/3572#discussion-diff-200991459R40

File: pypi_push.sh
Comment (r4vi) [reply pypi_push.sh]: @verdan then why not just run `yarn run build`?

**Comment by hydrosquall** [comment]
This is an exciting development that will make it easier to transpile all the javascript, right now the airflow UI does not load in IE11 due to the breaking usage of ES6 syntax. I've filed a new JIRA issue to tackle the javascript transpilation issue separately: 

https://issues.apache.org/jira/browse/AIRFLOW-2730

File: pypi_push.sh
Comment (verdan) [reply pypi_push.sh]: @r4vi that is absolutely possible. The primary manager I'm using is `npm` as it's pretty mature, has all the nice features like `npm audit`. And using `yarn` only to install the packages because yarn is pretty fast. One can use any of the yarn or npm for this.

File: pypi_push.sh
Comment (ashb) [reply pypi_push.sh]: The problem is there are other uploads than to PyPI.

File: airflow/www_rbac/static/package.json
Comment (ashb) [reply airflow/www_rbac/static/package.json]: They are not static asset files themselves though. Seeing this here I'd expect to be able to hit `http://my-airflow/static/package.json` or similar. Not having it under static removes that possible confusion entirely.

File: airflow/www_rbac/static/webpack.config.js
Comment (ashb) [reply airflow/www_rbac/static/webpack.config.js]: (... how can you have CSS code in a JS file?)

File: pypi_push.sh
Comment (ashb) [reply pypi_push.sh]: If we can get away with a single thing that would be cleaner.

File: pypi_push.sh
Comment (ashb) [reply pypi_push.sh]: Something like https://stackoverflow.com/questions/11331175/hook-to-add-commands-to-distutils-build is what I think we should do - and probably for `build`, which is a step of `sdist` I think.

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: https://github.com/webpack-contrib/mini-css-extract-plugin/tree/master/test/cases
P.S: I'm also using this plugin to combine multiple css files together in one minified css.

File: pypi_push.sh
Comment (verdan) [reply pypi_push.sh]: Okay, I'll update this to use only `npm` to eliminate the confusion.

Comment (ashb) [reply pypi_push.sh]: You can go for only Yarn if you'd rather. I don't mind which we use.

File: airflow/www_rbac/static/webpack.config.js
Comment (verdan) [reply airflow/www_rbac/static/webpack.config.js]: @ashb thoughts on this? ^

Comment (ashb) [reply airflow/www_rbac/static/webpack.config.js]: Not sure - no real strong feelings either way from me.

**Comment by verdan** [comment]
will re-open this PR after fixing all reviews and conflicts.

File: pypi_push.sh
Comment (verdan) [reply pypi_push.sh]: @ashb There's a catch behind this hack. I can override the commands that you mentioned in your comment, but then I need to do that for each command `build` and `sdist`. 
If I override `build` command, that would make it a new command and it won't get called automatically from other commands. 
One thing I can do is, introduce a new command `frontend` and then before each command, we'll need to run the new command. like: `python setup.py frontend sdist` something like that.

Comment (verdan) [reply pypi_push.sh]: If I override the main setup() function, and do the whole build process in that function, that would have other side effects. 
Users installing airflow using `pip install ..` would also run this command in that case, which is not required in any way. (and they'd have to install `npm` on their local systems as well)
The whole purpose of this build process is to make a ready to use build, with all processed Javascript and CSS files. For the initial release, can we do it with this script? as the PR is already pretty huge, and then we can create a new issue on JIRA to make this efficient.

Comment (ashb) [reply pypi_push.sh]: Having working assets from a `pip install .` sounds like a benefit to me. Though over-ridding `setup()` would be that running other commands (such as `setup.py --version` would take a long time, so that should be avoided)



I'm just checking something in the python sources, but you might be able to add a sub-command to build, or sdist that would do what we need without the over-load problems. Give me a moment.

Comment (ashb) [reply pypi_push.sh]: I think overriding the `build` command will work - they use https://github.com/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Lib/distutils/cmd.py#L292-L300 `get_finalized_command()` to call sub-commands, meaning it would call our build command, not the built in one.

Maybe.

A custom command in setup.py so we can do `setup.py compile_assets sdist` might be the right answer for now. I've not done a release myself yet so I don't know where the instructions on it are.

@bolkedebruin would changing the release to add an extra command before creating the sdist/bdist be okay? Where are the instructions so we can update them? Is it the wiki?

Comment (bolkedebruin) [reply pypi_push.sh]: Of course it is. @ashb

Comment (bolkedebruin) [reply pypi_push.sh]: Wiki indeed

**Comment by verdan** [comment]
Updated the description of this PR, added a new command in setup.py to compile assets, resolved all comments. 
Travis: https://travis-ci.org/apache/incubator-airflow/builds/402557114
@ashb ready for second review.

File: airflow/www_rbac/package.json
Comment (bolkedebruin) [inline airflow/www_rbac/package.json]: Apache Airflow

File: airflow/www_rbac/package.json
Comment (bolkedebruin) [inline airflow/www_rbac/package.json]: ??

2.0.0 ?

File: airflow/www_rbac/package.json
Comment (bolkedebruin) [inline airflow/www_rbac/package.json]: Apache Airflow

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Minor nits from my side. Please also make sure to check whether the license references in LICENSE are still correct and update if required.

File: airflow/www_rbac/package.json
Comment (bolkedebruin) [reply airflow/www_rbac/package.json]: Can we make this dynamic (ie. get it from version.py?)

Comment (ashb) [reply airflow/www_rbac/package.json]: It doesn't look like this can be dyanmic. We can set it with `npm version X` so it could be scripted. But given we aren't going to publish this to npm's repo do we need a version at all? If we do can we set it to "will-never-be-published" or similar?

Comment (ashb) [reply airflow/www_rbac/package.json]: From the [npm package.json docs](https://docs.npmjs.com/files/package.json):

> If you plan to publish your package, the most important things in your package.json are the name and version fields as they will be required.

That sounds like name and version could be left out in this case if we want.

Comment (verdan) [reply airflow/www_rbac/package.json]: `name` and `version` are totally optional if we don't want to publish this as an npm package anytime in the future. 
The version number can not be dynamic or set automatically, as this is JSON file. One need to update this manually. 
New npm packages can not have uppercase letters or spaces in the name. (it could be `apache-airflow` though)

Comment (ashb) [reply airflow/www_rbac/package.json]: It seems unlikely that we'd want to publish this -- its just the asset build pipeline, and I can't see a use case for someone else wanting to install this or depend upon it, so let's remove them.

Comment (verdan) [reply airflow/www_rbac/package.json]: 👍 that's true. will remove name and version from package.json in my next commit.

**Comment by verdan** [comment]
The travis failed because of following exception: `AirflowException: Pod Launching failed: Pod took too long to start`, do I need to restart the build or is this a known issue?

File: airflow/www_rbac/package.json
Comment (verdan) [reply airflow/www_rbac/package.json]: Fixed. Removed name and version number from package.json

**Comment by bolkedebruin** [comment]
Can you squash your commits @verdan ?

**Comment by bolkedebruin** [comment]
An fix the commit message so it adheres to the guidelines?

**Comment by verdan** [comment]
Squashed commits and fixed the commit message as per guidelines. Ready for the next review.

**Comment by verdan** [comment]
@ashb @bolkedebruin would you guys mind to take another look at this PR? or maybe even merge if everything looks fine now :)

**Comment by bolkedebruin** [comment]
merging. Any improvements can follow

**Comment by JacobHayes** [comment]
I don't see a separate issue, but the release process page still doesn't include the `compile_assets` release requirement. Should I open a new JIRA issue for that?

Relatedly, this makes `pip` installing Airflow from git a bit more difficult. Could `compile_assets` be added to `do_setup` to ease installation (and even the release process)?

**Comment by verdan** [comment]
@JacobHayes the release process page will be updated after the release of 1.10.x, as the current release-in-process is not using this `compile_assets` command yet. Having a JIRA issue for that is definitely a good idea to keep track of this change. 👍 
The discussion about having this in do_setup() can be followed here in the comments: https://github.com/apache/incubator-airflow/pull/3572#discussion_r200990849

We'll update the documentation on how to install Apache Airflow using pip from git directly. The JIRA issue to make that change is [AIRFLOW-2823](https://issues.apache.org/jira/browse/AIRFLOW-2823)

**Comment by JacobHayes** [comment]
[Added](https://issues.apache.org/jira/browse/AIRFLOW-2881)!

**Comment by r39132** [comment]
@verdan are you planning on updating the release process page now that 1.10.0 is out?

**Comment by verdan** [comment]
@r39132 I don't have access to update the release process page. I believe once this change will go live, then we'll need to update the release process page. If this is going to be released in the next release, then sure we need to update the release page.

**Comment by r39132** [comment]
@verdan You do now - CWIKI perms granted :-) 

Do you mean that you will wait until we retire the old UI before updating the release process? Isn't the rbac UI available already in the pypi releases? Kindly clarify? I believe the current process could be altered because rbac is already bundled with it with our releases.

**Comment by retornam** [comment]
@ashb the 1.10.4rc3  release https://github.com/apache/airflow/tree/1.10.4rc3/airflow/www/static and previous releases do not have the `dist` folder under the static dir, this is bad for users who pip install airflow because compile assets isn't called by default in `setup.py`. I think it will be best to include the generated dist folder in official releases to save users from having to install node

**Comment by ashb** [comment]
@retornam We don't check in the compiled assets - but do alreadd bundle them when making releases. If you `pip install apache-airflow` you'll get the compiled assets.

Note this only applies to the rbac UI. Classic UI didn't get this treatment

**Comment by retornam** [comment]
@ashb  this is from  freshly installed apache-airflow using pip

```├── ace.js
├── airflow.gif
├── bootstrap-theme.css
├── bootstrap-toggle.min.css
├── bootstrap-toggle.min.js
├── bootstrap3-typeahead.min.js
├── connection_form.js
├── d3.tip.v0.6.3.js
├── d3.v3.min.js
├── dagre-d3.js
├── dagre-d3.min.js
├── dagre.css
├── dataTables.bootstrap.css
├── favicon.ico
├── gantt-chart-d3v2.js
├── gantt.css
├── graph.css
├── jqClock.min.js
├── jquery.dataTables.css
├── jquery.dataTables.min.js
├── loading.gif
├── main.css
├── mode-sql.js
├── nv.d3.css
├── nv.d3.js
├── para
│   ├── parallel.css
│   ├── parallel.js
│   └── webgl-2d.js
├── pin.svg
├── pin_100.jpg
├── pin_100.png
├── pin_25.png
├── pin_30.png
├── pin_35.png
├── pin_40.png
├── pin_large.jpg
├── pin_large.png
├── screenshots
│   ├── gantt.png
│   ├── graph.png
│   └── tree.png
├── sort_asc.png
├── sort_both.png
├── sort_desc.png
├── theme-crimson_editor.js
├── tree.css
└── underscore.js
```

Some of the complied assets aren't there.
I'm also seeing `Please make sure to build the frontend in static/ directory and restart the server` in the webserver logs

**Comment by ashb** [comment]
Is that listing from www_rbac?

**Pull request closed without merging**
